### About this tree
This tree includes user-selected samples and contextual samples that are most closely related to those selected samples. This tree is designed to facilitate outbreak investigation, where typical questions include: are all of these samples actually part of the same outbreak? Did the outbreak originate from a single introduction event? Has an outbreak been contained to the localized setting or did it spill over and circulate within the wider community? For these types of investigations, it is critical to take into account the broader viral diversity of a region as a reference for the outbreak when making epidemiology inferences. In addition, this tree includes samples that are randomly selected over time across global regions to provide a more accurate estimate of the evolutionary rate.
Site numbering and genome structure uses [Wuhan-Hu-1/2019](https://www.ncbi.nlm.nih.gov/nuccore/MN908947) as reference. The phylogeny is rooted relative to early samples from Wuhan. Temporal resolution assumes a nucleotide substitution rate of 8 &times; 10^-4 subs per site per year. Mutational fitness is calculated using results from [Obermeyer et al (under review)](https://www.medrxiv.org/content/10.1101/2021.09.07.21263228v1). Full details on bioinformatic processing can be found [here](https://github.com/nextstrain/ncov).
