### About this tree
This tree is designed to provide an overview of samples collected only from within the user’s jurisdiction. This tree lacks contextual data that are crucial for making sound epidemiologic inferences, and thus we caution against its use for any purpose other than simply looking at viral diversity across the jurisdiction. These trees show all of the samples collected within a user’s jurisdiction by including data from CZ Gen Epi and GISAID, and if the total number of samples exceeds 2000, subsampling will occur and not all samples will be on the tree. 

Site numbering and genome structure uses [Wuhan-Hu-1/2019](https://www.ncbi.nlm.nih.gov/nuccore/MN908947) as reference. The phylogeny is rooted relative to early samples from Wuhan. Temporal resolution assumes a nucleotide substitution rate of 8 &times; 10^-4 subs per site per year. Mutational fitness is calculated using results from [Obermeyer et al (under review)](https://www.medrxiv.org/content/10.1101/2021.09.07.21263228v1). Full details on bioinformatic processing can be found [here](https://github.com/nextstrain/ncov).
