### About this tree
This tree is designed to provide a broad overview of the viral diversity in a user’s jurisdiction. It includes samples from the user's jurisdiction and samples from outside of it that are closely related to samples in the jurisdiction. These samples from outside of the jurisdiction provide important context for understanding and interpreting the viral diversity and its evolution within the jurisdiction. If more than 2000 samples are available from the user’s jurisdiction, random subsampling will occur.

Site numbering and genome structure uses [Wuhan-Hu-1/2019](https://www.ncbi.nlm.nih.gov/nuccore/MN908947) as reference. The phylogeny is rooted relative to early samples from Wuhan. Temporal resolution assumes a nucleotide substitution rate of 8 &times; 10^-4 subs per site per year. Mutational fitness is calculated using results from [Obermeyer et al (under review)](https://www.medrxiv.org/content/10.1101/2021.09.07.21263228v1). Full details on bioinformatic processing can be found [here](https://github.com/nextstrain/ncov).
