--- main_workflow.smk   2021-11-12 22:45:06.913872000 +0000
+++ main_workflow.smk.patched   2021-11-12 22:58:59.648973000 +0000
@@ -489,7 +489,7 @@
     benchmark:
         "benchmarks/priority_score_{build_name}_{focus}.txt"
     params:
-        crowding = 0.1,
+        crowding = config["crowding"]["crowding_penalty"],
         Nweight = 0.003
     conda: config["conda_environment"]
     shell:

